### PR TITLE
start writing di post

### PR DESCRIPTION
**Pull request by NullVoxPopuli**

### CODE REVIEW

File: content/2020-08-08-how-does-di-work.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+new to programming in general, or coming from another frontend ecosystem
+(especially React and Vue), is that they think Ember's dependency injection
+system is too complicated and magical --
+too hard to reason about or know where the injected services come from.
+I, too, was in that boat -- until I really dove into how it works -- it was
+then that I began to understand why dependency injection even exists, and
+how it's actually _simpler_ than _not_ having it at all.
+
+I can't speak to the Vue perspective, but professionally, I had my start
+to frontend development in React, and at the time there really only Redux and
```
Comment (lukemelia) [inline content/2020-08-08-how-does-di-work.md:30]:
```suggestion
to frontend development in React, and at the time there was really only Redux and
```

File: content/2020-08-08-how-does-di-work.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+MobX for state management -- but I only had the privilege of working with Redux
+and eventually React's Context Provider/Consumer pattern. There _is_ a little bit
+of overlap between React's Contexts and Ember's Services, but they differ in
+fundamental ways, and have pros and cons, which we'll get too.
```
Comment (lukemelia) [inline content/2020-08-08-how-does-di-work.md:34]:
```suggestion
fundamental ways, and have pros and cons, which we'll cover.
```

File: content/2020-08-08-how-does-di-work.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+1. A Service can be referenced from anywhere, regardless of where it is accessed
+2. A Service is a [singleton](https://en.wikipedia.org/wiki/Singleton_pattern)
+3. Services can reference each other (circular dependencies are valid)
+4. Access to the global namespace is not allowed
+
+
+
+This could be considered an ancestor to dependency injection, where there exists
+a shared `container` object in the module scope, still allowing for us to
+acheive the first three goals.
```
Comment (lukemelia) [inline content/2020-08-08-how-does-di-work.md:66]:
```suggestion
achieve the first three goals.
```

File: content/2020-08-08-how-does-di-work.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    let nextMove = this.container.ai.getMove();
+
+    this.container.ui.sendKeyPress(nextMove);
+  }
+}
+```
+
+This is a little better, we have abstracted away a bit of boilerplate, but there is still
+a "magic property", `container` -- this is generally where object oriented programming
+can get a negative reputation for -- a lock of _proper_ or _incomplete_ abstraction.
```
Comment (lukemelia) [inline content/2020-08-08-how-does-di-work.md:162]:
```suggestion
can get a negative reputation for -- a lack of _proper_ or _incomplete_ abstraction.
```

File: content/2020-08-08-how-does-di-work.md [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+
+But first, I'm a proponent of ["show, don't tell"](https://en.wikipedia.org/wiki/Show,_don%27t_tell),
+so we'll start by implementing dependency injection from scratch in order to demystify
+and have something concrete that can be referenced.
+
+## What is Dependency Injection?
+
+According to [Wikipedia](https://en.wikipedia.org/wiki/Dependency_injection)
+
+> *dependency injection* is a technique in which an object receives other obects that it depends on.
```
Comment (jrock2004) [inline content/2020-08-08-how-does-di-work.md:37]: Object is misspelled

File: content/2020-08-08-how-does-di-work.md [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+2. Testing is much easier and can be done in isolation
+
+For #1, there are many ways to share state between components, but I like that
+dependency injection provides a centralized pattern and location for that state
+as well as an ergonomic and light way to interact with that state.
+
+For #2, this is a little harder to boil down to a sentence or two, and ultimately
+comes down overall architecture of your app, how big your app is, and what sorts of
+things provide value when tested. For example, let's say you have some behavior
+for interacting with an external API, maybe it's the [Star Wars JSON api](https://stackoverflow.com/questions/14301389/why-does-one-use-dependency-injection),
```
Comment (jrock2004) [inline content/2020-08-08-how-does-di-work.md:68]: I was think the link was to show my the start wars json API, but its taking me to stackoverflow

Comment (NullVoxPopuli) [reply content/2020-08-08-how-does-di-work.md]: thanks!

File: content/2020-08-08-how-does-di-work.md (hunk 5 above)
Comment (NullVoxPopuli) [reply content/2020-08-08-how-does-di-work.md]: thanks!

**Pull request merged**